docs/processing.md

```diff
@@ -0,0 +1,289 @@
+- Navigate to a path within the telemetry to operate on it
+- Define an operation, and possibly operation arguments
+
+We can try to model these into a query language, in particular allowing the first two points to be shared among all
+processing operations, and only have implementation of individual types of processing need to implement operators that
+the user can use within an expression.
+
+Telemetry is modeled in the collector as [`pdata`](https://github.com/open-telemetry/opentelemetry-collector/tree/main/model/pdata)
+which is roughly a 1:1 mapping of the [OTLP protocol](https://github.com/open-telemetry/opentelemetry-proto/tree/main/opentelemetry/proto).
+This data can be navigated using field expressions, which are fields within the protocol separated by dots. For example,
+the status message of a span is `status.message`. A map lookup can include the key as a string, for example `attributes["http.status_code"]`.
+
+Operations can be scoped to the type of a signal (`span`, `metric`, `log`), with all of the flattened points of that
+signal being part of a query space. Virtual fields are added to access data from a higher level before flattening, for
+`resource`, `library_info`. For metrics, the structure presented for processing is actual data points, e.g. `NumberDataPoint`, 
+`HistogramDataPoint`, with the information from higher levels like `Metric` or the data type available as virtual fields.
+
+Navigation can then be used with a simple expression language for identifying telemetry to operate on.
+
+`... where name = "GET /cats"`
+`... from span where attributes["http.target"] = "/health"`
+`... where resource.attributes["deployment"] = "canary"`
+`... from metric where descriptor.metric_type = gauge`
+`... from metric where descriptor.metric_name = "http.active_requests"`
+
+Fields should always be fully specified - for example `attributes` refers to the `attributes` field in the telemetry, not
+the `resource`. In the future, we may allow shorthand for accessing scoped information that is not ambiguous.
+
+Having selected telemetry to operate on, any needed operations can be defined as functions. Known useful functions should
+be implemented within the collector itself, provide registration from extension modules to allow customization with
+contrib components, and in the future can even allow user plugins possibly through WASM, similar to work in 
+[HTTP proxies](https://github.com/proxy-wasm/spec). The arguments to operations will primarily be field expressions,
+allowing the operation to mutate telemetry as needed.
+
+### Examples
+
+Remove a forbidden attribute such as `http.request.header.authorization` from all telemetry.
+
+`delete(attributes["http.request.header.authorization"])`
+
+Remove a forbidden attribute from spans only
+
+`delete(attributes["http.request.header.authorization"]) from span`
+
+Remove all attributes except for some
+
+`keep(attributes, "http.method", "http.status_code") from metric`
```

Comment:
We call the "signals" as [traces/metrics/logs] may want to be consistent here and use plural at least, and probably use "traces" instead of "spans"?